enhancements/20240611-vm-schedule-backup.md

```diff
@@ -0,0 +1,173 @@
+# Support VM Schedule Backup
+
+## Summary
+
+This feature supports scheduled backup of the VM periodically.
+
+### Related Issues
+
+- https://github.com/harvester/harvester/issues/2756
+
+## Motivation
+
+### Goals
+
+Harvester supports backup of the VM on a scheduled basis with the configuration option to keep a specific number of backups. 
+
+The Harvester controller will keep updating a list of VM backups even if the VM is not running, and in the meantime, clean up the outdated backups along with the Longhorn snapshots to guarantee space usage efficiency.
+
+The Harvester controller will suspend the schedule when the number of failed backups reaches a pre-defined (configurable) threshold. After making sure the backup target is normal, the user can ask to resume the scheduled backup, then the Harvester controller will check the remote target's health, and resume the scheduled job if certain conditions are met.
+
+## Proposal
+
+### User Stories
+
+Although [Longhorn provides recurring backup](https://longhorn.io/docs/1.6.2/snapshots-and-backups/scheduling-backups-and-snapshots/) that is per volume and not VM-based, so it brings the challenge to restore the VM from these volume backups.
+
+With this feature, the user can have a **list of VM backups** which are kept renovated according to the format of a [CRON expression](https://en.wikipedia.org/wiki/Cron#CRON_expression), and the user also can choose anyone from the list to restore the VM.
+
+
+### API changes
+
+## Design
+
+### Implementation Overview
+
+#### CRD
+
+- Add new resource `ScheduleVMBackup` to control VM schedule backup jobs
+```
+type ScheduleVMBackupSpec struct {
+	// +kubebuilder:validation:Required
+	Cron string `json:"cron"`
+
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:=8
+	// +kubebuilder:validation:Maximum=254
+	// +kubebuilder:validation:Minimum=2
+	Retain int `json:"retain"`
+
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:=4
+	// +kubebuilder:validation:Minimum=2
+	MaxFailure int `json:"maxFailure"`
```

Comment:
Since we use `ConsecutiveFailure` to determine whether to suspend the cronjob, do we want to change `MaxFailure` to `MaxConsecutiveFailure`? I think it's more clear for users to know when the cronjob will be suspended.

Reply:
good idea, thanks @FrankYang0529

Reply:
Hi @FrankYang0529 

After discussing with @ibrokethecloud, I changed `status.failure` to count the consecutive failure backups since the last successful one, since the #total failure backups could be obtained from `status.vmbackupInfo`, and it will not affect the controller's behavior, thanks.